### PR TITLE
Complex containers

### LINKED ISSUE #1116: Complex containers
This is a tracking issue for adding `complex_matrix`, `complex_vector`, and `complex_row_vector` support to the language.

### PR DESCRIPTION
**Pull request by WardBrian**
Closes #1116 . This branch has gotten to a really good place. I believe with the exception of https://github.com/stan-dev/math/issues/2671 that the basic tests I've written here should compile and pass.

Thanks to previous PRs like #1102, #1115, #1106, most of these changes are pretty "boring", mostly just doing the obvious things based on the existing container types and the promotion semantics.

Things this PR adds:
1. `complex_matrix`, `complex_vector`, `complex_row_vector` types. The behavior and code generation of these largely mirrors the existing Eigen types.
2. Promotion from `matrix`, etc. to `complex_matrix` as necessary (e.g. in assignment, in function calls)
3. Basic math signatures (`add`, `subtract`, `multiply`, `minus`, `elt_multiply`, `elt_divide`) and basic container functions (e.g `rows` and the like)

Things this PR does not add:
1. Many, many other function signatures which would be valid or interesting for complex types. There are things like `pow`, which might work already, and things like FFTs, which will need math support. I leave these all for later PRs.
2. Add "mixed" signatures for things like `multiply: (complex_vector,  row_vector)=> complex_matrix`. NB: _this does currently work_, but it does so by promoting the `row_vector` to `complex_row_vector`. We can later change this for speedups. 

#### Submission Checklist

- [x] Run unit tests
- Documentation
    - [ ] If a user-facing facing change was made, the documentation issue is: https://github.com/stan-dev/docs/issues/433

## Release notes

Added complex containers: `complex_matrix`, `complex_vector`, and `complex_row_vector` with basic arithmetic support. More special functions and specializations will follow. Note that real-valued containers will be promoted to their complex variants as needed

## Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the BSD 3-clause license (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: src/stan_math_backend/Stan_math_code_gen.ml [hunk 1]
```diff
@@ -44,6 +44,7 @@ let pp_templates ~defaults ppf templates =
   match templates with
   | [] -> ()
   | _ ->
+      (* TODO collate requires into one require_all *)
```
Comment (WardBrian) [inline src/stan_math_backend/Stan_math_code_gen.ml:47]: This is a note to remember to come back and do something like what @SteveBronder was working on in #1107. I took a slightly different approach in this PR, but the same method of combining them into one `require_all` should work. My first priority was to get it compiling as-is, however.

**Comment by SteveBronder** [comment]
@WardBrian just put up the PR! Are you able to make a local cmdstan and try out https://github.com/stan-dev/math/pull/2680?

**Comment by WardBrian** [comment]
@SteveBronder I ran this branch with the pr here: https://jenkins.flatironinstitute.org/blue/organizations/jenkins/Stan%2FStanc3/detail/PR-1133/7/pipeline/ and everything is green!

I would still like to see https://github.com/stan-dev/math/issues/2676 get addressed whenever @rok-cesnovar has the chance before this would get merged, just to automatically test all the declared complex signatures.

**Comment by SteveBronder** [comment]
Will look at this tmrw!

**Comment by WardBrian** [comment]
@SteveBronder did you get a chance to look at this?

File: test/integration/good/code-gen/complex_numbers/basic_ops_mix.stan [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  row_vector[N] vrowvec;
+  real v;
+}
+
+
+transformed parameters {
+  complex_matrix[N,N] gq_c_matrix;
+
+  // matrix-matrix multiply and elt
+  gq_c_matrix = cmat * cvmat;
```
Comment (SteveBronder) [inline test/integration/good/code-gen/complex_numbers/basic_ops_mix.stan:31]: Is there a reason these are labeled gq?

**Review by SteveBronder — CHANGES_REQUESTED** [review]
Everything looks good to me! Though I think we need to do one thing in `mem_pattern` to check that any matrix or vector that is assigned to a complex matrix or vector is demoted to `AoS`. Besides that I'd like to try to cleanup the requires a bit on the functions so we don't have so many as I worry that could hurt compile time a good bit. But I can't do that then this PR looks good

File: src/stan_math_backend/Stan_math_code_gen.ml [hunk 3]
```diff
@@ -101,24 +102,38 @@ let maybe_templated_arg_types (args : Program.fun_arg_decl) =
...
-      | Some t -> Some (Require (require_for_arg a, t))
-      | None -> None )
+let require_templates (names : string option list) (args : Program.fun_arg_decl)
+    =
+  let require_for_arg arg t =
+    let requires arg =
+      match arg with
+      | UnsizedType.URowVector ->
+          [ Require ("stan::require_row_vector_t", t)
+          ; Require ("stan::require_not_vt_complex", t) ]
```
Comment (SteveBronder) [inline src/stan_math_backend/Stan_math_code_gen.ml:112]: I think to put these all in one require we want to

1. Make all of these just logic checks like `is_row_vector` and replace `require_not_vt_complex` and `is_real_value_type`
2. concatenate them all into one `require_all_t<>`. So a check for some like the signature below would have the associated templates

```stan
matrix foo(complex_vector a, matrix b)
```

Would be something like 

```cpp
template <typename T1__, typename T2__, require_all_t<is_eigen_vector<T1__>, is_complex_vt<T1__>, is_eigen_matrix_dynamic<T2__>, is_real_vt<T2__>>* = nullptr>
```

Ftr I'm fine with it as is until we add that to the math library

File: test/integration/good/code-gen/complex_numbers/basic_ops_mix.stan (hunk 2 above)
Comment (WardBrian) [reply test/integration/good/code-gen/complex_numbers/basic_ops_mix.stan:31]: I copied/pasted and didn't change the names. I'll do it now

File: src/stan_math_backend/Stan_math_code_gen.ml (hunk 3 above)
Comment (WardBrian) [reply src/stan_math_backend/Stan_math_code_gen.ml:112]: Do we have all the required template machinery to do this right now? I think that the `require_(not_)vt_complex` stuff is all generated by macros right?

File: src/frontend/Typechecker.ml [hunk 4]
```diff
@@ -1400,18 +1416,28 @@ and check_sizedtype cf tenv sizedty =
...
+      SMatrix (mem_pattern, te1, te2)
+  | SComplexVector e ->
+      let te = check e "complex vector sizes" in
+      SComplexVector te
+  | SComplexRowVector e ->
+      let te = check e "complex row vector sizes" in
+      SComplexRowVector te
+  | SComplexMatrix (e1, e2) ->
+      let te1 = check e1 "Complex matrix sizes" in
+      let te2 = check e2 "Complex matrix sizes" in
```
[the remark this answers is not in the document]
Comment (WardBrian) [reply src/frontend/Typechecker.ml:1435]: We currently say "Matrix sizes" for the equivalent check for real matrices. We can swap both over if you want, I think that would make sense

**Comment by SteveBronder** [comment]
This all looks good! I think it just needs the thing in mem_pattern then it's good to merge

**Comment by WardBrian** [comment]
I'm not sure I understand what needs to change in mem_pattern to get the necessary behavior

**Comment by SteveBronder** [comment]
I'll take a look today. It's just that on the first pass we want to make sure that if a matrix/vector is being assigned to a complex matrix/vector then it is tagged as not possible to be an `SoA` matrix. So in `query_initial_demotable_stmt` we want to look for assignments whose meta type is `ComplexMatrix/Vector` and if the rhs expression is an `Autodiffable` `Matrix/Vector` then we want to add any named matrix/vector objects in the rhs expression to the list of matrices that cannot be promoted to `SoA` matrices. So we want to make sure the below on O1 will not have `A_p` be an `SoA` matrix.

```stan
parameters {
 matrix[10, 10] A_p;
}
transformed parameters {
 complex_matrix[10, 10] A_complex_tp = A_p;
}

https://github.com/stan-dev/stanc3/blob/master/src/analysis_and_optimization/Mem_pattern.ml#L357

**Comment by WardBrian** [comment]
I'd appreciate you taking a look, that part of the code is tough for me

**Comment by SteveBronder** [comment]
See https://github.com/stan-dev/stanc3/pull/1133/commits/bb816a57934e74f8bd54b64e2ce8e157ce2200f9 I think that's all it needs so that it knows that is an object that cannot be promoted. Do you still need to add any new error messages specifically for complex types?

**Comment by WardBrian** [comment]
I believe all the error messages should be good to go

**Comment by WardBrian** [comment]
Looks like it needs a `dune promote`

**Review by SteveBronder — APPROVED** [review]
lgtm!

**Comment by WardBrian** [comment]
Thanks @SteveBronder - I'll hit merge on this as soon as 2.29.2 is out and good to go

**Comment by WardBrian** [comment]
Well, also after @bob-carpenter and I update the relevant docs

**Pull request merged**